L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc

```diff
@@ -0,0 +1,150 @@
+
+#include "DataFormats/L1TParticleFlow/interface/PFJet.h"
+#include "DataFormats/JetReco/interface/Jet.h"
+#include "DataFormats/L1TParticleFlow/interface/PFCandidate.h"
+#include "L1Trigger/Phase2L1ParticleFlow/interface/BJetId.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+
+#include "DataFormats/L1TCorrelator/interface/TkPrimaryVertex.h"
+#include "DataFormats/L1Trigger/interface/Vertex.h"
+#include "DataFormats/L1Trigger/interface/VertexWord.h"
+
+#include "DataFormats/Math/interface/deltaR.h"
+#include <cmath>
+#include <vector>
+
+using namespace l1t;
+
+class L1BJetProducer : public edm::stream::EDProducer<edm::GlobalCache<BJetTFCache>> {
+public:
+  explicit L1BJetProducer(const edm::ParameterSet&, const BJetTFCache*);
+  ~L1BJetProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  static std::unique_ptr<BJetTFCache> initializeGlobalCache(const edm::ParameterSet&);
+  static void globalEndJob(const BJetTFCache*);
+
+private:
+  std::unique_ptr<BJetId> fBJetId_;
+  void produce(edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+
+  edm::EDGetTokenT<edm::View<l1t::PFJet>> jets_;
+  bool fUseRawPt_;
+  double fMinPt_;
+  double fMaxEta_;
+  unsigned int fMaxJets_;
+  int fNParticles_;
+  edm::EDGetTokenT<std::vector<l1t::TkPrimaryVertex>> fTkVtx_;
+  edm::EDGetTokenT<std::vector<l1t::VertexWord>> fVtxEmu_;
+  bool fVtxEmulation_;
+  edm::EDGetTokenT<l1t::PFCandidateCollection> fL1PFToken_;
+};
+
+static constexpr float track_trigger_eta_max = 2.5;
+
+L1BJetProducer::L1BJetProducer(const edm::ParameterSet& cfg, const BJetTFCache* cache)
+    : jets_(consumes<edm::View<l1t::PFJet>>(cfg.getParameter<edm::InputTag>("jets"))),
+      fUseRawPt_(cfg.getParameter<bool>("useRawPt")),
+      fMinPt_(cfg.getParameter<double>("minPt")),
+      fMaxEta_(cfg.getParameter<double>("maxEta")),
+      fMaxJets_(cfg.getParameter<int>("maxJets")),
+      fNParticles_(cfg.getParameter<int>("nParticles")),
+      fVtxEmulation_(cfg.getParameter<bool>("vtxEmulation")),
+      fL1PFToken_(consumes<l1t::PFCandidateCollection>(cfg.getParameter<edm::InputTag>("L1PFObjects"))) {
+  std::string lNNFile = cfg.getParameter<std::string>("NNFileName"); 
+  fBJetId_ = std::make_unique<BJetId>(
+      lNNFile.find("qModel") == std::string::npos ? "input:0" : "Conv1D_1_input:0", cache, lNNFile, fNParticles_);
+  if (fVtxEmulation_) {
+    fVtxEmu_ = consumes<std::vector<l1t::VertexWord>>(cfg.getParameter<edm::InputTag>("vtx"));
+  } else {
+    fTkVtx_ = consumes<std::vector<l1t::TkPrimaryVertex>>(cfg.getParameter<edm::InputTag>("vtx"));
+  }
+  produces<edm::ValueMap<float>>("L1PFBJets");
+}
+std::unique_ptr<BJetTFCache> L1BJetProducer::initializeGlobalCache(const edm::ParameterSet& cfg) {
+  tensorflow::setLogging("3");
+  std::string lNNFile = cfg.getParameter<std::string>("NNFileName");
+  edm::FileInPath fp(lNNFile);
+  BJetTFCache* cache = new BJetTFCache();
```

Comment:
in order to get it into central CMSSW this has to be replaced with
`auto cache = std::make_unique<BJetTFCache>()`
(and you might have to `return std::move(cache)`, I'm not 100% sure, but the compiler will tell)

Also, wouldn't it make more sense to have the constructor of BJetTFCache take the path to the graph, calling `loadGraphDef` internally, and own the pointer to the graph. The pointer never needs to be updated during the job, so it doesn't need to be an atomic, so it can be a `unique_ptr` and you avoid to have a destructor

Reply:
Ok yes, thanks for the suggestion. I've modified things accordingly.